<launch>
  <!-- General arguments -->
  <arg name="param_file"
       default="$(find wavemap_2d_ros)/config/wavemap_server.yaml"
       doc="Name of the file from which to read wavemap's server params."/>
  <arg name="debug_mode"
       default="false"
       doc="Run the mapper in a debugger (gdb)."/>
  <arg name="show_rviz"
       default="true"
       doc="Automatically launch Rviz."/>

  <!-- Dataset params -->
  <arg name="pointcloud_topic"
       doc="Name of the pointcloud topic to read from."/>
  <arg name="rosbag_path"
       doc="Replay the rosbag at the specified path."/>
  <!-- NOTE: Relative paths will be resolved w.r.t. ROS_HOME (default ~/.ros) -->

  <!-- Wavemap params -->
  <arg name="data_structure_type"
       default="dense_grid"/>
  <arg name="measurement_model_type"
       default="beam_model"/>

  <!-- Process the rosbag with wavemap as fast as possible -->
  <param name="/use_sim_time" value="true"/>
  <group>
    <arg if="$(arg debug_mode)" name="launch_prefix"
         value="gdb -ex run --args"/>
    <arg unless="$(arg debug_mode)" name="launch_prefix" value=""/>

    <node name="wavemap_2d" pkg="wavemap_2d_ros"
          type="wavemap_2d_rosbag_processor"
          output="screen" args="-alsologtostderr" required="true"
          launch-prefix="$(arg launch_prefix)">
      <rosparam file="$(arg param_file)"/>
      <param name="pointcloud_topic"
             value="$(arg pointcloud_topic)"/>
      <param name="rosbag_path" value="$(arg rosbag_path)"/>
      <param name="data_structure_type" value="$(arg data_structure_type)"/>
      <param name="measurement_model_type"
             value="$(arg measurement_model_type)"/>
    </node>
  </group>

  <!-- Show Rviz -->
  <group if="$(arg show_rviz)">
    <arg name="rviz_file"
         default="$(find wavemap_2d_ros)/config/wavemap_server.rviz"/>
    <node pkg="rviz" type="rviz" name="$(anon rviz)" output="screen"
          args="-d $(arg rviz_file)"/>
  </group>
</launch>
